CDAP-6399 Create ns command cli

lumig242 opened the pull request:
Enhance the create command of CLI.

Take an extra json of properties to construct the namespace config. 

Example usage here:

`create namespace ns1 decription "hive-database=db1 hbase-namespace=habsens1 principal=Booshan keytab-path=/path/to/keytab"`

JIRA: https://issues.cask.co/browse/CDAP-6399
buid: http://builds.cask.co/browse/CDAP-DUT4586

lumig242 committed:
CDAP-6399 Enhance namespace create CLI command to allow passing a namespace config

anwar6953 commented:
can you give an example of usage?

lumig242 commented:
@anwar6953  Added an example of usage above. Basically the format is consistent with the json properties passed when creating a dataset instance.

anwar6953 commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/CreateNamespaceCommand.java:
```diff
@@ -35,6 +37,12 @@
 
   private final NamespaceClient namespaceClient;
+  private static final String NAMESPACE_PRINCIPAL = "principal";
+  private static final String NAMESPACE_KEYTAB_PATH = "keytab-path";
```
shouldn't these keys be consistent with the keys used, when submitting the json via curl?

lumig242 replied:
Oh yes. Did not realized this point. Changed the key names.

gokulavasan commented:
@lumig242 Can description have spaces in it?

lumig242 commented:
@gokulavasan  Yes. But user should wrap it with double quotations., like this:
`create namespace ns "description with space" "key=value"`

bdmogal commented:
@lumig242 because how how the namespace config is accepted, we cannot autocomplete things like hive-database, hbase-namespace, principal, keytabpath, etc, right? We should take them as separate parameters, so it is not necessary to remember these keywords.

bdmogal commented:
e.g. `create namespace ns1 with [description <description] [root-directory <root-directory>] [hbase-namespace <hbase-namespace>] [hive-database <hive-database>] [principal <principal>] [keytab-path <keytab-path>]`.

Because the literals/keywords (hbase-namespace, hive-database, etc) already understood by the command in this case, user can easily autocomplete them using the `<TAB>` key. Also, having these makes the arguments non-positional, so they can be entered in any order, based upon what you want to enter.

bdmogal commented:
See examples of this in the grant/revoke commands.

lumig242 commented:
@bdmogal @anwar6953  Changed the format as described above.

bdmogal commented on cdap-cli-tests/src/test/java/co/cask/cdap/cli/CLIMainTest.java:
```diff
@@ -666,11 +666,11 @@ private static void testNamespacesOutput(CLI cli, String command, final List<Nam
     PrintStream output = new PrintStream(outputStream);
     Table table = Table.builder()
-      .setHeader("name", "description")
+      .setHeader("name", "description", "configs")
```
we should also split this into multiple columns right?

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/ArgumentName.java:
```diff
@@ -78,6 +78,12 @@
 
   NAMESPACE_NAME("namespace-name"),
   NAMESPACE_DESCRIPTION("namespace-description"),
```
should be called just description?

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/CreateNamespaceCommand.java:
```diff
@@ -45,22 +45,38 @@ public CreateNamespaceCommand(CLIConfig cliConfig, NamespaceClient namespaceClie
+    String keytabPath = arguments.getOptional(ArgumentName.NAMESPACE_KEYTAB_PATH.toString(), null);
+    String hbaseNamespace = arguments.getOptional(ArgumentName.NAMESPACE_HBASE_NAMESPACE.toString(), null);
+    String hiveDatabase = arguments.getOptional(ArgumentName.NAMESPACE_HIVE_DATABASE.toString(), null);
+    String schedulerQueueName = arguments.getOptional(ArgumentName.NAMESPACE_SCHEDULER_QUEUENAME.toString(), "");
```
why default this to ""?

lumig242 replied:
When constructing the namespaceConfig, there is a if statement there to check, `if schedulerQueueName == null` then `schedulerQueueName = ""` to make sure it is not null.

anwar6953 replied:
we can still default this to null, and it will be set to empty string when build() is called below

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/CreateNamespaceCommand.java:
```diff
@@ -45,22 +45,38 @@ public CreateNamespaceCommand(CLIConfig cliConfig, NamespaceClient namespaceClie
     String name = arguments.get(ArgumentName.NAMESPACE_NAME.toString());
-    String description = arguments.get(ArgumentName.NAMESPACE_DESCRIPTION.toString(), "");
+
+    String description = arguments.getOptional(ArgumentName.NAMESPACE_DESCRIPTION.toString(), "");
```
why default this to ""?

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/DescribeNamespaceCommand.java:
```diff
@@ -53,11 +53,11 @@ public void perform(Arguments arguments, PrintStream output) throws Exception {
     NamespaceMeta namespaceMeta = namespaceClient.get(namespace);
     Table table = Table.builder()
-      .setHeader("name", "description")
+      .setHeader("name", "description", "configs")
```
let's split into multiple columns instead of showing a json

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/ListNamespacesCommand.java:
```diff
@@ -47,11 +47,11 @@ public ListNamespacesCommand(CLIConfig cliConfig, NamespaceClient namespaceClien
   public void perform(Arguments arguments, PrintStream output) throws Exception {
     Table table = Table.builder()
-      .setHeader("name", "description")
+      .setHeader("name", "description", "configs")
```
ditto

bdmogal commented:
@lumig242 minor comments. Other than that, can you add some tests for custom mapping in `CLIMainTest#testNamespaces`?

lumig242 commented:
Addressed comments. Here is how the current CLI `list namespaces` command looks like.

![image](https://cloud.githubusercontent.com/assets/4253578/17415193/9385fbb6-5a3d-11e6-8cd6-c862f2b7e865.png)

bdmogal commented on cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java:
```diff
@@ -121,4 +121,35 @@ public String toString() {
+   *
+   * @return the String to print
+   */
+  public String prettyPrintCLI() {
```
If this is only used in CLI, let's move it to the CLI classes. classes in cdap-proto are public, so we can't make incompatible changes here later. Should not add CLI specific code here

bdmogal commented on cdap-cli-tests/src/test/java/co/cask/cdap/cli/CLIMainTest.java:
```diff
@@ -450,6 +450,10 @@ public void testPreferences() throws Exception {
+    final String keytab = "keytab";
+    final String principal = "principal";
+    final String hbaseNamespace = "hbase";
+    final String schedulerQueueName = "queue";
```
should add hive too?

anwar6953 commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/NamespaceCommandUtils.java:
```diff
@@ -0,0 +1,59 @@
+    if (namespaceConfig.getKeytabURI() != null) {
+      builder.append("keytab-URI='").append(namespaceConfig.getKeytabURI()).append("', ");
+    }
+    if (builder.length() > 0) {
```
add a comment that this removes the final ", ".

bdmogal commented:
LGTM
